### ISSUE TITLE
Can I modify/edit a value manually in redux-devtools without using a reducer?

### ISSUE THREAD

**Issue by aguynamedben** [zalmoxisus/redux-devtools-extension#502]
It seems like I might be missing something very simple, but I couldn't figure this out.

I'm building an Electron app, and after the app launches, in order to debug some business logic, I want to modify a single value in a single piece of my state (account -> access_token). I want to replace an OAuth access token with a known stale access token so that I can test that the refresh_token is used correctly.

Is there a way to modify state manually with redux devtools? I've learned that I can use Chrome Devtools (Application Tab) to modify localStorage directly (I'm using redux-persist), but that is painful.

It seems everybody says "use reducers to modify state", which makes sense from an architecture standpoint, but it seems like the whole point of a redux devtools is to allow me to tinker and inspect (similar to how you can edit HTML/CSS from Chrome Devtools).

What I tried

1. Open redux-devtools
2. In the Inspector view, switch to the State tab to see entire state
3. Within that, click the Raw tab. That brings up a text editor for the state.
4. Manually modify the state.
5. I don't see how to save the state? The Persist button below doesn't save it. Anywhere I click from here seems to not save the state and instead resets it to its previous value.

Thanks for the time you spend working on this tool. It's *extremely* value to my team. Sorry if PEBKAC.

**Comment by TomGroombridge** [comment]
Agreed it would be a great feature if you could manipulate the state directly.

**Comment by augbog** [comment]
I know this hasn't gotten a lot of feedback but personally I don't think this is a good idea? I believe redux devtools does the right thing and requires you to dispatch actions and then rely on your reducer to edit the state which is what the core principles of Redux want you to enforce. Manipulating the state directly has very weird implications and you may be testing application states that your application can never hit.

In the case from the original OP, 

> I want to modify a single value in a single piece of my state (account -> access_token).

How does that `access_token` get set in the first place? I imagine some dispatch action needs to take place. In that case, dispatch that same action and let your reducer handle it for you.

**Comment by philihp** [comment]
This would be very powerful.

Consider that you're an engineer working on a presentational component that could receive a certain state, while in parallel another engineer is working on the action/reducer that produces that state. Being able to mock that state would unblock you from waiting for the action/reducer be complete.

**Comment by mlg87** [comment]
@tmaiaroto add it to chrome and it will detect stores

**Comment by alekline** [comment]
This is great, thank you so much!

Please note that in my app, the case code is as follows:


```
import { fromJS, mergeDeep, set } from 'immutable';
    ....
/* develblock:start */
case 'HACK_STATE': {
  console.log('Testing hack_state')
  const mergeValue = (action as any).state || (action as any).value;
  const mergeHack = (action as any).path ? set({}, (action as any).path, mergeValue) : mergeValue;
  return mergeDeep(state, fromJS(mergeHack));
}
/* develblock:end */
```

And then, use this in the Action Dispatcher of NgRx Store DevToos
```
{
  type: 'HACK_STATE',
  path: 'hcgov.applicationId',
  value: 1234
}
```

**Comment by meymeynard** [comment]
any update on this feature?

**Comment by zalmoxisus** [comment]
I'll implement it after 3.0, when will finish extension's rewriting and new UI. Additionally to states, the plan is also to edit dispatched actions, changing the payloads. And if, there's interest, duplicating actions; so after editing would be 2 options to overwrite the current one (and recompute states) or dispatch as a new action.

**Comment by danielo515** [comment]
That sounds very cool @zalmoxisus, do you know when it is going to be released?

**Comment by dfhincapiem** [comment]
Great to hear news from this new amazing feature!!!!

**Comment by cjz9032** [comment]
Great to hear news from this new amazing feature!!!!

**Comment by danielo515** [comment]
Oh, yea yes, it is `Great to hear news from this new amazing feature!!!!`, but without knowing when it is going to be released or if it is released already it is useless.

**Comment by Chris-Kim** [comment]
Great to hear news from this new amazing feature!!!!

**Comment by ss-gnalvesteffer** [comment]
Great to hear news from this new amazing feature!!!!

**Comment by dakipro** [comment]
Hello, any update on this feature? It would help quite a lot when developing, especially handling edge "hypothetical" cases

**Comment by aliankarali** [comment]
Please give this issue the attention it deserves, even beofre these UI updates or whatnot.

**Comment by lklepner** [comment]
@zalmoxisus Can you share any updates on the timing of this feature?

**Comment by kamok** [comment]
Coming from Vuex Dev Tools, this feature would be a massive quality of life improvement for x Redux developers. I will try to respond to the arguments that Redux Dev Tools enforces "proper" developer practice by forcing you to dispatch actions. 

Some apps that I work with directly feeds API responses into Redux to keep it as a single source of truth. (eventhough the app does NOT directly manipulate some values, eg readonly data). When this happens, it's very troublesome to have to manipulate the backend `development` data. Or worse, if you do not have access to it. 

In tests you can throw some mock data into the store and run them. But you will not be able to get screenshots and actual visual confirmation (so I can put them into my PR for acceptance).

**Comment by Tzinov15** [comment]
Any updates on a release date for this?

**Comment by ropinheiro** [comment]
I was just trying to edit state in Redux Dev Tools, then though that it shouldn't be possible at all, because of the "only through reducers" rule, then I thought "hey, but it's Dev Tools, to help devs debug and test faster, not to be production", then I found this page and noticed that I'm not alone on those thoughts... for more than 2 years (late... but not least :P) Soooo, when will this very useful feature be available? :)

**Comment by VishiNehra** [comment]
pleajh

**Comment by lance-cfa** [comment]
It's been over 2 years since this update was suggested. Amazing to see how quickly the 3 participants in vue-devtools got this done.

https://github.com/vuejs/vue-devtools/issues/699
ShaneHudson commented on Jun 28, 2018 (proposed feature)
@Akryum Akryum added feature request priority: medium accepted proposition labels on Jul 25, 2018
@Akryum Akryum added this to TODO in v4.x on Jul 30, 2018
@Akryum Akryum moved this from TODO to In Progress in v4.x on Aug 1, 2018
@hootlex hootlex mentioned this issue on Aug 1, 2018

@Akryum - have any time to implement this for Redux dev tools?

**Comment by aguynamedben** [comment]
Hey @lance-cfa from Chick-fil-A I'm sure they'd accept a pull request if you sent one

**Comment by matheuslino** [comment]
@zalmoxisus, any update on this feature? It's been almost 3 years since the first comment =/

**Comment by PremRanjanDev** [comment]
Editing of redux state would be a great feature. 
For example, right now I am getting some expression from API call which has to be evaluated at the client-side. I need to check my evaluation logic works for different expressions. Simply I cannot edit the state (API response value) in redux dev-tool and see the effect on UI. And yes, my project does not have the implementation of mock data, mostly it relies on dev-server API calls.

**Comment by sandersiim** [comment]
Have to agree with all the previous comments, for some reason I assumed editing state directly is possible in Redux DevTools, and then I found this thread for my disappointment...
This would be an invaluable tool for debugging and development!

**Comment by SeanDez** [comment]
This is an essential feature, please add it :)

**Comment by AlbinCederblad** [comment]
Any updates on this?

**Comment by robmarshall** [comment]
This would be super helpful.

**Comment by mmulhearn** [comment]
I hope this is still planned; it would be great for dev

**Comment by aswinpchn** [comment]
I think that allowing the user complete access to modify the Redux store will be a security. So, that in a production environment Redux Devtools should not be allowed.

**Comment by rufat** [comment]
Still waiting. It would be great for debugging.

**Comment by djwashburn** [comment]
seriously? the dev tools don't allow editing the state live? what's the point then?

**Comment by danielo515** [comment]
> seriously? the dev tools don't allow editing the state live? what's the point then?

all the other features they offer

**Comment by AndyOooh** [comment]
Was this ever implemented?

**Comment by Trollyal** [comment]
> Was this ever implemented?

Does not look like. As a workaround, you can export the state, modify the json outside and reimport it back. Tedious, but works.

**Comment by RyanPWalker** [comment]
.

**Comment by themarcel** [comment]
Bumping this in 2025!
